app/proc/processor.go

```diff
@@ -32,31 +32,43 @@ type Processor struct {
 }
 
 // Do activate loop of goroutine for each feed, concurrency limited by p.Conf.Concurrent
-func (p *Processor) Do() {
+func (p *Processor) Do(ctx context.Context) error {
 	log.Printf("[INFO] activate processor, feeds=%d, %+v", len(p.Conf.Feeds), p.Conf)
 
 	for {
-		swg := syncs.NewSizedGroup(p.Conf.System.Concurrent, syncs.Preemptive)
-		for name, fm := range p.Conf.Feeds {
-			for _, src := range fm.Sources {
-				name, src, fm := name, src, fm
-				swg.Go(func(context.Context) {
-					p.processFeed(name, src.URL, fm.TelegramChannel, p.Conf.System.MaxItems, fm.Filter)
-				})
-			}
-			// keep up to MaxKeepInDB items in bucket
-			if removed, err := p.Store.removeOld(name, p.Conf.System.MaxKeepInDB); err == nil {
-				if removed > 0 {
-					log.Printf("[DEBUG] removed %d from %s", removed, name)
-				}
-			} else {
-				log.Printf("[WARN] failed to remove, %v", err)
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			p.processFeeds(ctx)
+		}
+	}
+}
+
+func (p *Processor) processFeeds(ctx context.Context) {
+	log.Printf("[DEBUG] refresh started")
+	swg := syncs.NewSizedGroup(p.Conf.System.Concurrent, syncs.Preemptive, syncs.Context(ctx))
+	for name, fm := range p.Conf.Feeds {
+		for _, src := range fm.Sources {
+			name, src, fm := name, src, fm
+			swg.Go(func(context.Context) {
+				p.processFeed(name, src.URL, fm.TelegramChannel, p.Conf.System.MaxItems, fm.Filter)
+			})
+		}
+	}
+	swg.Wait()
```

Comment:
I see what you mean, but don't really recall. This won't cause any issues anyway, as bolt itself "syncing" on writes/deletes, but to me, the logical place to put this logic would be inside of swg.Go, or maybe add it to processFeed

Reply:
I think a possible issue here is the case when `removeOld` finishes faster than one of `processFeed` goroutines and some items are added to the db after removal loop. In this case we can have more than `MaxKeepInDB` items in the db when we exit `processFeeds`.
It's not a big issue and these items will be removed by the next `removeOld` call anyway but we can have unstable results in the test, can't we?

So I put this logic to `processFeed`, this will fix the issue and be more logical place for this code.